### ISSUE TITLE
Type issue with custom matcher

### ISSUE THREAD

**Issue by shannonrothe** [molefrog/wouter#180]
I'm running into an issue implementing a custom matcher. I've followed the example in the README but TS complains about the type.

The `makeRegexp` function passed to `makeCachedMatcher` (or `makeMatcher`) is expecting a return of type `RegExp`, but the example outlines `{ keys: Key[], regexp: RegExp }`.

```jsx
const convertPathToRegexp = (path: string) => {
  let keys: Key[] = [];

  const regexp = pathToRegexp(path, keys, { strict: true });
  return { keys, regexp };
};

const strictMatcher = makeCachedMatcher(convertPathToRegexp);
```

results in

```
Argument of type '(path: string) => { keys: Key[]; regexp: RegExp; }' is not assignable to parameter of type '(pattern: string, keys?: { name: string | number; }[] | undefined) => RegExp'.
  Type '{ keys: Key[]; regexp: RegExp; }' is missing the following properties from type 'RegExp': exec, test, source, global, and 13 more.ts(2345)
```

Am I missing something here, or is this an issue? I'm happy to raise a PR if it's as simple as changing the return type of `makeRegexp` 👍🏼 

Thanks for your work on this package too! It's really great and a nice light alternative to `react-router` with a flexible API.

**Comment by shannonrothe** [comment]
FWIW I can navigate the type issue but I still get console errors:

```jsx
const strictMatcher = makeCachedMatcher((path: string) => {
  const { regexp } = convertPathToRegexp(path);
  return regexp;
});
```

![image](https://user-images.githubusercontent.com/803013/109932520-69a76100-7d1e-11eb-8924-26a424317cd2.png)

I think it gets confused by my custom `ProtectedRoute` component:

```jsx
<Router matcher={strictMatcher}>
  <Switch>
    <Route path="/login" component={Login} />
    <ProtectedRoute path="/logout" component={Logout} />
    <Route path="/register" component={Register} />

    {/* Specific document */}
    <ProtectedRoute path="/:slug/documents/:id" component={Document} />

    <Layout>
      {/* List all root folders */}
      <ProtectedRoute path="/" component={Home} />
      <ProtectedRoute path="/teams" component={Teams} />

      {/* Sub-folders */}
      <ProtectedRoute path="/:slug" component={Folder} />
    </Layout>
  </Switch>
</Router>
```

which looks like this:

```jsx
export const ProtectedRoute = (props: RouteProps) => {
  const {
    userStore: { isAuthenticated },
  } = useStores();

  return isAuthenticated ? <Route {...props} /> : <Redirect to="/login" />;
};
```

**Comment by cbbfcd** [comment]
```js
export default function makeMatcher(
  makeRegexpFn?: (
    pattern: string,
    keys?: Array<{ name: string | number }>
  ) => RegExp // here.
): MatcherFn;

PR welcome, 😂
```

**Comment by molefrog** [comment]
Released in v2.7.4
